idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/purity/PurityPlugin.kt

```diff
@@ -2,39 +2,73 @@ package arrow.meta.ide.plugins.purity
 import org.jetbrains.kotlin.lexer.KtTokens
 import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.KtProperty
 import org.jetbrains.kotlin.resolve.calls.tower.isSynthesized
-import org.jetbrains.kotlin.util.ReturnsCheck
+import org.jetbrains.kotlin.types.KotlinType
 
 val IdeMetaPlugin.purity: Plugin
   get() = "PurityPlugin" {
     meta(
       addApplicableInspection(
-        defaultFixText = "Suspend",
+        defaultFixText = "SuspendFunction",
         inspectionHighlightType = { ProblemHighlightType.ERROR },
         kClass = KtNamedFunction::class.java,
         highlightingRange = { f -> f.nameIdentifierTextRangeInThis() },
-        inspectionText = { f -> "Function should be suspended" },
+        inspectionText = { f -> "Function: ${f.name} should be suspended" },
         applyTo = { f, project, editor ->
           f.addModifier(KtTokens.SUSPEND_KEYWORD)
         },
         isApplicable = { f: KtNamedFunction ->
           f.nameIdentifier != null && !f.hasModifier(KtTokens.SUSPEND_KEYWORD) &&
             f.resolveToDescriptorIfAny()?.run {
-              !isSuspend && !isSynthesized && !isSuspendLambdaOrLocalFunction() && (ReturnsCheck.ReturnsUnit.check(this) || ExtendedReturnsCheck.ReturnsNothing.check(this)
-                || ExtendedReturnsCheck.ReturnsNullableNothing.check(this))
+              !isSuspend && !isSynthesized && !isSuspendLambdaOrLocalFunction() &&
+                returns(f) { impureTypes }
             } == true
         },
         level = HighlightDisplayLevel.ERROR,
         groupPath = ArrowPath + "PurityPlugin"
+      ),
+      addApplicableInspection(
+        defaultFixText = "ImpureProperty",
+        inspectionHighlightType = { ProblemHighlightType.ERROR },
+        kClass = KtProperty::class.java,
+        highlightingRange = { prop -> prop.nameIdentifierTextRangeInThis() },
+        inspectionText = { prop -> "Property: ${prop.name} has an impure initializer" },
+        applyTo = { prop, project, editor ->
+          // TODO: Find a refactoring strategy which works for some cases; for now just warn the user
+          /*prop.initializer?.let { body: KtExpression ->
+            modify(body) { b: KtExpression ->
+              b.safeAs<KtLambdaExpression>()?.let {
+                createExpressionByPattern("$1$0", it.functionLiteral, createIdentifier("suspend"))
+              } ?: createExpressionByPattern("suspend { $0 }", b)
+            }
+          }*/
+        },
+        isApplicable = { prop: KtProperty ->
+          prop.resolveToDescriptorIfAny()?.run {
+            !returns(
+              f = { t: KotlinType -> t.constructor },
+              types = { suspendedFunctionTypes }
+            ) && returns(prop) { impureTypes }
+          } == true
+        },
+        level = HighlightDisplayLevel.ERROR,
+        groupPath = ArrowPath + "PurityPlugin"
       )
     )
-  }
\ No newline at end of file
+  }
+
+private val KotlinBuiltIns.impureTypes: List<KotlinType>
```

Comment:
This is probably extremely slow given these properties are continuously computed and it looks like they could just be a `val` in the plugin that get initialized the moment you can access the `KotlinBuiltIns`

Reply:
How can we improve that ?